rust/projects/networking/project.md

```diff
@@ -0,0 +1,288 @@
+etc. Both are extensible, supporting different backends, for logging to the
+console, logging to file, logging to the system log, etc.
+
+[`log`]: https://docs.rs/log/
+[`slog`]: https://docs.rs/slog/
+
+The major difference is that `log` is fairly simple, logging only formatted
+strings; `slog` is feature-rich, and supports "structured logging", where log
+entries are typed and serialized in easily-parsed formats.
+
+`log` dates from the very earliest days of Rust, where it was part of the
+compiler, then part of the standard library, and finally released as its own
+crate. It is maintained by the Rust Project. `slog` is newer and maintained
+independently. Both are widely used.
+
+For both systems, one needs to select a "sink" crate, one that the logger
+sends logs to for display or storage.
+
+_Read about both of them, choose the one that appeals to you, add them as
+dependencies, then modify `kvs-server` to initialize logging on startup, prior
+to command-line parsing._ Set it up to output to stderr (sending the logs
+elsewhere additionally is fine, but they must go to stderr to pass the tests in
+this project).
+
+On startup log the server's version number. Also log the configuration. For now
+that means the IP address and port, and the name of the storage engine.
+
+
+## Part 3: Client-server networking setup
+
+Next we're going to set up the networking. For this project you are going to be
+using the basic TCP/IP networking APIs in `std::net`: `TcpListener` and
+`TcpStream`. Before thinking about the protocol, modify `kvs-server` to listen
+for and accept connections, and `kvs-client` to initiate connections.
+
+For this project, the server is synchronous and single-threaded. That means that
+you will listen on a socket, then accept connections, and execute and respond to
+commands one at a time. In the future we will re-visit this decision multiple
+times on our journey toward an asynchronous, multi-threaded, and
+high-performance database.
+
+Thank about your manual testing workflow. Now that there are two executables to
+deal with, you'll need a way to run them both at the same time. If you are like
+many, you will use two terminals, running `cargo run --bin --tikv-server` in
+one, where it runs until you press CTRL-D`, and `cargo run --bin --tikv-client`
+in the other.
+
+This is a good opportunity to use the logging macros for debugging. Go ahead and
+log information about every accepted connection.
+
+
+## Part 4: Implementing commands across the network
+
+Now it's time to implement the key/value store over the network, remotely
+executing commands that until now have been implemented within a single process.
+As with the file I/O you worked on in the last project to create the log, you
+will be serializing and streaming commands with the `Read` and `Write` traits.
+
+You are going to design a network protocol.
+
+TODO: Can't think of much to say about this
```

Comment:
Maybe point them to the HTTP/1.1 examples on Wikipedia? Implementing super minimal HTTP/1.1 is very easy and I imagine many will do so anyways.